### PR TITLE
BUG: `GroupBy.quantile` implicitly sorts `index.levels`

### LINKED ISSUE #53009: BUG: SeriesGroupby with sort=False behaves inconsistently with .quantiles(...).unstack()
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd

ind = pd.MultiIndex.from_tuples([(0, 'a', 'B'),
                                 (0, 'a', 'A'),
                                 (0, 'b', 'B'),
                                 (0, 'b', 'A'),
                                 (1, 'a', 'B'),
                                 (1, 'a', 'A'),
                                 (1, 'b', 'B'),
                                 (1, 'b', 'A'),
                                 ], names=['sample', 'cat0', 'cat1'])
ds = pd.Series(range(8), index=ind)
print(ds)

s01 = ds.groupby(level=['cat0', 'cat1'], sort=False).quantile([0.2, 0.8])
print('\n(cat0, cat1) Quantiles As Series: Result as expected:\n', s01)
d01 = s01.unstack()
print('\nQuantiles Unstacked: Result as expected:\n', d01)

s1 = ds.groupby(level='cat1', sort=False).quantile([0.2, 0.8])
print('\ncat1 Quantiles As Series: Result as expected:\n', s1)
d1 = s1.unstack()
print('\nQuantiles Unstacked: NOT as expected; cat1 index SORTED alphabetically:\n', d1)

print('\nQuantiles Unstacked, by iterrows:')
for (cat, q) in d1.iterrows():
    print(f'{cat}: {q.values}')
```

### Issue Description

I need to preserve the pre-defined order of MultiIndex labels in both 'cat0' and 'cat1', when calculating quantiles. The problem is that the result is as expected when the index 'cat1' in the example variable d01 is the second of two groupby levels, but does NOT give a similar expected result when 'cat1' is the single groupby level, as in the variable d1 in the example.

### Expected Behavior

In my code example, I expected the unstacked DataFrame d1 to retain the index order ['B', 'A'], same as in the Series variable s1, and in analogy to the similarly unstack-ed DataFrame variable d01. 

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 37ea63d540fd27274cad6585082c91b1283f963d
python           : 3.10.7.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 22.4.0
Version          : Darwin Kernel Version 22.4.0: Mon Mar  6 21:01:02 PST 2023; root:xnu-8796.101.5~3/RELEASE_ARM64_T8112
machine          : arm64
processor        : arm
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.UTF-8

pandas           : 2.0.1
numpy            : 1.23.3
pytz             : 2022.2.1
dateutil         : 2.8.2
setuptools       : 63.2.0
pip              : 23.1.2
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.6.2
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : 3.0.10
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.9.1
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : 2023.3
qtpy             : None
pyqt5            : None

</details>

### PR DESCRIPTION
**Pull request by Charlie-XIAO**
- [x] Closes #53009
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests)
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit)
- [x] Added an entry in the latest `doc/source/whatsnew/v2.1.0.rst` file

### CODE REVIEW

File: pandas/tests/groupby/test_quantile.py [hunk 1]
```diff
@@ -471,3 +471,27 @@ def test_groupby_quantile_dt64tz_period():
...
+            (1, "a", "B"),
+            (1, "a", "A"),
+            (1, "b", "B"),
+            (1, "b", "A"),
+        ],
+        names=["sample", "cat0", "cat1"],
+    )
+    ds = pd.Series(range(8), index=ind)
+
+    s1 = ds.groupby(level="cat1", sort=False).quantile([0.2, 0.8])
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_quantile.py:493]: Just for style consistency in the tests, can you rename `ds -> ser` and `s1 -> result`

File: pandas/tests/groupby/test_quantile.py [hunk 2]
```diff
@@ -471,3 +471,27 @@ def test_groupby_quantile_dt64tz_period():
...
+
+    # Check that levels are not sorted
+    expected = pd.core.indexes.frozen.FrozenList([["B", "A"], [0.2, 0.8]])
+    tm.assert_equal(s1.index.levels, expected)
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_quantile.py:497]: Please check the entire object `s1 / result` (not just `index.levels`). If it makes it easier, it is okay to reduce the size of the data `ds / ser` above, or change the values.

File: doc/source/whatsnew/v2.1.0.rst [hunk 3]
```diff
@@ -401,6 +401,7 @@ Groupby/resample/rolling
   the function operated on the whole index rather than each element of the index. (:issue:`51979`)
 - Bug in :meth:`DataFrameGroupBy.apply` causing an error to be raised when the input :class:`DataFrame` was subset as a :class:`DataFrame` after groupby (``[['a']]`` and not ``['a']``) and the given callable returned :class:`Series` that were not all indexed the same. (:issue:`52444`)
 - Bug in :meth:`GroupBy.groups` with a datetime key in conjunction with another key produced incorrect number of group keys (:issue:`51158`)
+- Bug in :meth:`GroupBy.quantile` may implicitly sort ``index.levels`` (:issue:`53009`)
```
Comment (rhshadrach) [inline doc/source/whatsnew/v2.1.0.rst:404]: Can you add `with ``sort=False``. Also, ` replace `index.levels` doesn't seem clear to me, what do you think of replacing this with `the result index`?

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the PR! Lookng good - some small requests.

**Comment by Charlie-XIAO** [comment]
Hi @rhshadrach, thanks for your review! I've updated the changelog and the tests as you suggested. However, I'm not sure if `index.levels` is being compared. I tried using
```python
qind = pd.MultiIndex(
    [["A", "B"], [0.2, 0.8]],
    [[1, 1, 0, 0], [0, 1, 0, 1]],
    names=["cat1", None],
)
```
and it still passed the test. Is there any recommended way of fixing this?

**Comment by rhshadrach** [comment]
Thanks for identifying this @Charlie-XIAO - I opened #53095. For this PR, I think it'd be good to keep the assert_series_equal, and assert the index levels as you had it.

**Comment by Charlie-XIAO** [comment]
Thanks @rhshadrach! I've retrieved the index levels check.

File: pandas/core/groupby/groupby.py [hunk 4]
```diff
@@ -4317,5 +4317,8 @@ def _insert_quantile_level(idx: Index, qs: npt.NDArray[np.float64]) -> MultiInde
         codes = [np.repeat(x, nqs) for x in idx.codes] + [np.tile(lev_codes, len(idx))]
         mi = MultiIndex(levels=levels, codes=codes, names=idx.names + [None])
     else:
-        mi = MultiIndex.from_product([idx, qs])
+        lev_codes, lev = Index(qs).factorize()
+        levels = [idx, lev]
+        codes = [np.repeat(range(len(idx)), nqs), np.tile(lev_codes, len(idx))]
+        mi = MultiIndex(levels=levels, codes=codes, names=[idx.name, None])
```
Comment (rhshadrach) [inline pandas/core/groupby/groupby.py:4323]: I'm seeing a bit of a performance regression here from the use of 64 bit ints in codes. The current path will use int8, int16, int32 when appropriate, resulting in 50% speedup in this op (~200ms on my machine) with the following example:

```
size = 10000
idx = pd.Index(range(size), name='cat1')
qs = np.array([e / size for e in range(size)])
```

Converting the inputs of `np.repeat` and `np.tile` to `int16` in this example performs slightly better than main. I think we should find a way to factorize this using the appropriate number of bits for the dtype, but I'll have to take a deeper look to best figure out how.

Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:4323]: @Charlie-XIAO this avoids the perf regression:

```
from pandas.core.dtypes.cast import coerce_indexer_dtype

nidx = len(idx)
nqs = len(qs)
lev_codes, lev = Index(qs).factorize()
lev_codes = coerce_indexer_dtype(lev_codes, lev)
idx_codes = coerce_indexer_dtype(np.arange(nidx), idx)
levels = [idx, lev]
codes = [np.repeat(idx_codes, nqs), np.tile(lev_codes, nidx)]
mi = MultiIndex(levels=levels, codes=codes, names=[idx.name, None])
```

The import, if necessary, should happen at the top of the file.

Comment (Charlie-XIAO) [reply pandas/core/groupby/groupby.py:4323]: @rhshadrach Thank you so much! I will commit these changes ASAP.

**Comment by Charlie-XIAO** [comment]
Thanks for your review @rhshadrach! I've pushed the changes that avoid performance regression when `idx` is not `MultiIndex` as you suggested. Please let me know if similar changes should apply to the case when `idx` is `MultiIndex`.

**Comment by rhshadrach** [comment]
> Please let me know if similar changes should apply to the case when `idx` is `MultiIndex`.

Yes! It looks to me like these two blocks can now be largely combined.

**Review by rhshadrach — APPROVED** [review]
lgtm

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @Charlie-XIAO

**Pull request merged**